sources/standard/filesystem/__init__.py

```diff
@@ -0,0 +1,178 @@
+    modified_at: dlt.sources.incremental[pendulum.DateTime] = dlt.sources.incremental(
+        "mtime",
+        initial_value=DEFAULT_START_DATE,
+        allow_external_schedulers=True,
+    ),
+    chunksize: int = DEFAULT_CHUNK_SIZE,
+) -> TDataItem:
+    """List the files available in the filesystem.
+    
+    Args:
+        fs_client (AbstractFileSystem): The filesystem client.
+        bucket_url (str): The url to the bucket.
+        filename_filter (Optional[str], optional): The filter to apply to the files in glob format.
+        modified_at (dlt.sources.incremental[pendulum.DateTime], optional): The date to start collecting files.
+        chunksize (int, optional): The number of files to process at once.
+
+    Returns:
+        TDataItem: The list of files.
+    """
+
+    if not filename_filter:
+        filename_filter = "*"
+    files_url = os.path.join(bucket_url, filename_filter)
+    files = fs_client.glob(files_url, detail=True)
+
+    last_modified = modified_at.last_value
+
+    file_list = []
+    for file, md in files.items():
+        for dt_field in ["created", "mtime"]:
+            md[dt_field] = ensure_pendulum_datetime(md[dt_field])
+        # Skip files modified before last run
+        if last_modified and md["mtime"] < last_modified:
+            continue
+        md["path"] = file
+        md["content_type"] = mimetypes.guess_type(file)[0]
+        file_list.append(md)
+
+    for i in range(0, len(file_list), chunksize):
+        yield file_list[i : i + chunksize]
+
+
+@dlt.transformer(
+    name="files",
+    write_disposition="merge",
+    merge_key="file_name",
+    primary_key="file_name",
+)
+def copy_files(
+    items: TDataItems,
+    fs_client: AbstractFileSystem,
+    storage_path: str = STORAGE_PATH,
+) -> TDataItem:
+    """Reads files and copy them to local directory.
+    
+    Args:
+        items (TDataItems): The list of files to copy.
+        fs_client (AbstractFileSystem): The filesystem client.
+        storage_path (str, optional): The path to store the files.
+
+    Returns:
+        TDataItem: The list of files copied.
+    """
+    for file in items:
+        file_path = file["path"]
+        filename = os.path.basename(file_path)
+        target_path = os.path.join(storage_path, filename)
+        os.makedirs(os.path.dirname(target_path), exist_ok=True)
+        fs_client.cp(file_path, target_path)
+
+        file_hash = fs_client.checksum(file_path)
+        modified = fs_client.modified(file_path)
+
+        file_md = FileModel(
+            file_name=filename,
+            file_path=os.path.abspath(target_path),
+            content_type=file["content_type"],
+            modification_date=ensure_pendulum_datetime(modified),
+            data_hash=file_hash,
+        )
+
+        yield file_md.dict()
+
+
+@dlt.transformer(
+    name="content",
+    write_disposition="merge",
+    merge_key=None,
+    primary_key=None,
+)
+def extract_files(
```

Comment:
this is a part of the demo of the filesystem pipeline! and we have one problem here: there's no clear way to pass `fs_client` - you can't expect that instance will be automatically created.

my take is that we subclass Python built-in dict and add a special property that will hold `credentials`. we'll also implement a helper function that when passed such dict, will use `credentials` to get `AbstractFilesystem`. I wrote that in my ticket #216